[Draft] DR-0001 Architecture clarifications (#748)
blindnet-io/product-management: decisions/DR-0001-architecture-clarifications.md

| **Author** | Amaury Rousseau (amaury@blindnet.io)                              |

## Context and Problem Statement

The [High Level Conceptualization](https://github.com/blindnet-io/product-management/blob/d3dad802ce52fb1488cbda8302d3c36c94c3b47b/refs/high-level-conceptualization/README.md),
[High Level Architecture](https://github.com/blindnet-io/product-management/blob/d3dad802ce52fb1488cbda8302d3c36c94c3b47b/refs/high-level-architecture/readme.md),
and the [Architecture specification](https://github.com/blindnet-io/product-management/blob/ae344653217b03d3857fba63f5d716e43c72263b/specifications/architecture/README.md)
leave some free space for interpretation when implementing the devkit components.
To enable an efficient, coordinated and coherent development of all components, some choices have to be made.

## Considered Options

This record is the result of multiple meetings to discuss this matter between @filipblnt, @m4rk055 and @TheKinrar.

## Decision Outcome

### Interaction of Data Capture and Storage components

The Storage component only acts as an abstract interface to any backing storage implementation.

> It is able to store data as byte streams, which might be short text representations, like Data Capture metadata,

Reviewer: ```suggestion
It allows to link actual data records with PRIV metadata corresponding to it. It also allows to interpret actions resulting from Privacy Computation Engine, intro instructions (e.g. SQL queries) executable over data stroage.
```
After our discussion today, I think we are limiting the role of Storage to only an interface to link with other part of the architecture, so I suggest this change

or large files, like the actual data of a Capture Fragment.
File paths are used to designate data when reading and writing.
It is agnostic of the format of the data, and does not know either if the data is encrypted or not.

> The Data Capture component is responsible for initiating data captures, but also for later retrieval of the data.

Reviewer: ```suggestion
The Data Capture component is responsible for producing PRIV metadata, but also for identifing the corresponding data recrds for later retrieval of the data.
```

Therefore, it stores the minimal set of metadata needed for these operations, which includes IDs of captures, fragments
and users, fragment selectors and permissions.
When a client app is creating a Data Capture, the Data Capture Component is responsible for generating the ID of the
capture and of its fragments. It will return these to the app, alongside the necessary information for uploading the
actual data to the Storage Component. This includes a signed token authorizing the user for the upload process.
Fragment data never transits through the Data Capture Fragment.

### Encryption

> Encryption being optional, all components other than the Encryption Component need to behave in the same way, whether

Reviewer: A System might be configured to follow different [response scenarios](https://github.com/blindnet-io/product-management/blob/devkit-schemas/refs/schemas/priv/scenarios.md#response) when data is encrypted and when it isn't. I.e. it may be needed to allow Nested Response, for certain types of requests, when data is encrypted, and disallow it (and force Direct Response) when it isn't encrypted.

Author: I'll fix the wording. Here we meant specifically the other components mentioned in this DR, that is Data Capture and Storage. Privacy Compiler is not included indeed.

the data is encrypted or not. The Data Capture and Storage component do not need to communicate with the
Encryption Component, only the client app does.
The Encryption component is responsible for storage of all encryption keys, including user asymmetric keys and fragment

> symmetric keys. Fragment data is the only data to be encrypted.

Reviewer: It seems to me that some keys (and not only fragment data) may also end up being encrypted by other keys.

Reviewer 2: Fragment symmetric keys are encrypted with users' public keys.

### Authentication and authorization

> All components depend on the Identity Component for authentication purposes. The Identity Component will typically

Author: > The Submitter MAY be anonymous at capture time.

The idea is that an anonymous submitter would still use a temporary user ID for submitting the data, because we still need to authenticate that they are allowed to submit.

take an Authorization header as input, and output a verified user ID, and some additional user info if needed.

Each component performs authorization of requests based on the output of the Identity Component.

### Versioning

Versioning is handled by the Data Capture component. Each version of a fragment has its own ID which is used to store
and retrieve the data in the Storage component. When a new version of a fragment is created, a new version ID is
generated and the new data can be uploaded to the storage.

This requires a change in the
[devkit Data Model](https://github.com/blindnet-io/product-management/tree/devkit-dbmodel/specifications/model):
Capture Fragment now has Versions, which have UUIDs too, and the Data value is moved from the fragment to the version.

> 

Reviewer: ```suggestion

### Privacy Request Capture Interface

In Q3 2022, we are not implementing the Global scope of Privacy Request Capture Interface, but only the local scope.

```
